### ISSUE TITLE
Redirect to different HTTPS host broken

### ISSUE THREAD

**Issue by hasso** [prometheus/blackbox_exporter#237]
### Host operating system: output of `uname -a`

Linux probe 4.9.50 #1 SMP Thu Sep 14 08:57:58 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2630 v2 @ 2.60GHz GenuineIntel GNU/Linux

### blackbox_exporter version: output of `blackbox_exporter -version`

blackbox_exporter, version 0.9.1 (branch: HEAD, revision: 2272b4e637a591168d5b24a3943e15ac15ecba80)
  build user:       root@27fc71dcd45b
  build date:       20170921-13:19:29
  go version:       go1.8.3

### What is the blackbox.yml module config.
```
  http_2xx_ip4:
    prober: http
    timeout: 5s
    http:
      preferred_ip_protocol: "ip4"
```

### Log snippet
```
ts=2017-09-26T11:30:56.950294668Z caller=main.go:115 module=http_2xx_ip4 target=http://foo.com level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-09-26T11:30:56.950378987Z caller=utils.go:43 module=http_2xx_ip4 target=http://foo.com level=info msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-09-26T11:30:56.950762868Z caller=utils.go:63 module=http_2xx_ip4 target=http://foo.com level=info msg="Resolved target address" ip=127.16.1.5
ts=2017-09-26T11:30:56.950794289Z caller=http.go:190 module=http_2xx_ip4 target=http://foo.com level=info msg="Making HTTP request" url=http://[127.16.1.5] host=foo.com
ts=2017-09-26T11:30:56.951863609Z caller=http.go:150 module=http_2xx_ip4 target=http://foo.com level=info msg="Received redirect" url=https://bar.com
ts=2017-09-26T11:30:56.954761741Z caller=http.go:194 module=http_2xx_ip4 target=http://foo.com level=error msg="Error for HTTP request" err="Get https://bar.com: x509: certificate is valid for www.bar.com, bar.com, not foo.com"
ts=2017-09-26T11:30:56.954796702Z caller=main.go:128 module=http_2xx_ip4 target=http://foo.com level=error msg="Probe failed" duration_seconds=0.004475863
```
### What did you expect to see?

I'd expect a probe to succeed.

### What did you see instead?

Probe fails.

**Comment by brian-brazil** [comment]
This will be a little tricky to handle, we'd need to make the RoundTripper do different things on different requests.

**Comment by pavelhritonenko** [comment]
I have a similar issue when `http://mysite.com` site redirects the request to `http://www.mysite.com` and it redirects to `https://www.mysite.com`. It seems that last probe doesn't use SNI because I received error that certificate is not valid for default ingress certificate:

```Logs for the probe:
ts=2017-10-15T08:16:08.493433681Z caller=main.go:115 module=http_2xx target=http://uptick.com level=info msg="Beginning probe" probe=http timeout_seconds=9.5
ts=2017-10-15T08:16:08.493603593Z caller=utils.go:43 module=http_2xx target=http://uptick.com level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-15T08:16:08.513232577Z caller=utils.go:50 module=http_2xx target=http://uptick.com level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address uptick.com: no suitable address found"
ts=2017-10-15T08:16:08.516235448Z caller=utils.go:63 module=http_2xx target=http://uptick.com level=info msg="Resolved target address" ip=35.158.184.245
ts=2017-10-15T08:16:08.51627591Z caller=http.go:190 module=http_2xx target=http://uptick.com level=info msg="Making HTTP request" url=http://[35.158.184.245] host=uptick.com
ts=2017-10-15T08:16:08.518510844Z caller=http.go:150 module=http_2xx target=http://uptick.com level=info msg="Received redirect" url=http://www.uptick.com/
ts=2017-10-15T08:16:08.552600395Z caller=http.go:150 module=http_2xx target=http://uptick.com level=info msg="Received redirect" url=https://www.uptick.com/
ts=2017-10-15T08:16:08.560485506Z caller=http.go:194 module=http_2xx target=http://uptick.com level=error msg="Error for HTTP request" err="Get https://www.uptick.com/: x509: certificate is valid for ingress.local, not uptick.com"
ts=2017-10-15T08:16:08.560530479Z caller=main.go:128 module=http_2xx target=http://uptick.com level=error msg="Probe failed" duration_seconds=0.067022724```

**Comment by germansokolov13** [comment]
I thought I'd make Prometheus to monitor uptime of the Internet connection of the server. I made blackbox_exporter use http to `https://google.com` and to my huge dismay it won't work. On the other hand `https://stackoverflow.com` works. But `content_length` metric yields `-1` for SO. Yes, a negative value of one.
How is it possible? What was the module tested against? And how do I approach monitoring Internet connection if even just requesting `google.com` won't work.

**Comment by brian-brazil** [comment]
Please don't ask general support questions on bugs, the user mailing list is a better place for that.

**Comment by germansokolov13** [comment]
I thought it was related to this bug. Sorry if I'm wrong.

**Comment by crepetl** [comment]
Hi,
Any updates about this issue?
From my really low understanding of the issue, would it be interesting to process redirects inside Blackbox exporter instead of using RoundTripper behavior?
I honestly did not check for changes on net/http.

**Comment by SuperQ** [comment]
We also recently ran into this trying to upgrade from 0.8.1 to 0.12. Under normal use, I would expect `https://` redirects to work correctly.

@discordianfish Do you have any ideas here on how to make `RoundTripper ` work correctly with hostname change redirects on https?

**Comment by kamusin** [comment]
another one bite the dust here.

**Comment by 1medusa1** [comment]
Hello, i am using the blackbox exporter provided by prometheus.io you don't need to install or make the go application as it is already there ready to execute, i am also having the same issue with redirects, but this is normal if it is looking for 2xx status code and found 400 then it should fail, but there must be an option to specify redirects on blackbox.yml or prometheus.yml ?

Is this possible, i didn't do a go build :s

**Comment by jacksontj** [comment]
I ran into this issue, and spent some time digging into it. The issue is actually a combination of (1) trying to hit a specific IP (2) setting ServerName and (3) prom's config mechanism for http clients.


Fundamentally the issue begins [here](https://github.com/prometheus/blackbox_exporter/blob/d47e1ee5560c36dbe7c77af3ecafe624ef1c7cfb/prober/http.go#L222) where the ServerName is set. "Normally" this wouldn't need to be set as go's http stdlib actually uses the name from the URL as the Servername -- and this is updated on redirects. Unfortunately blackbox exporter is actually doing the [DNS lookup *for* the HTTP client](https://github.com/prometheus/blackbox_exporter/blob/master/prober/http.go#L211) -- this was done to [force the ip protocol](https://github.com/prometheus/blackbox_exporter/pull/382). Now that the DNS resolution is done up front the ServerName *must* be set in the TLS config otherwise the initial request will set the ServerName to the resolved IP (which likely isn't in the cert).

**TLDR; feature to force IP resolution requires that the ServerName be set**

So I see a few ways to solve this:

##  Update the ServerName in `client.CheckRedirect`.
The http client has a method (CheckRedirect) which is called before executing any redirect. In this method we could do something like:
```
client.CheckRedirect = func(r *http.Request, via []*http.Request) error {
	tlsConfig.ServerName = r.URL.Host
}
```
to set the ServerName to the domain name of the redirect.
I tested that this approach works in testing, but has a few problems in blackbox exporter:

1. These redirects don't force an IP version -- this is actually an issue with this *today* but this would perpetuate the problem
2. There is no access to the tlsConfig from the blackbox exporter. The call to [config.NewHTTPClientFromConfig](https://github.com/prometheus/blackbox_exporter/blob/master/vendor/github.com/prometheus/common/config/http_config.go#L120) returns a client with no pointers to the TLSConfig; so we can't update it

## override the DialContext to force a network type
The main issue with this seems to be that the pre-resolution of the name to IP is causing us to *require* setting a ServerName. In addition other situations (e.g. redirect) still fetch using the default 6 then 4 lookups that go does by default. If we instead wrapped the dial method for the HTTPClient then we could force DNS resolution to ipv4 or ipv6 as we desire, which means we'd (1) not require setting the ServerName in the TLS config -- which fixes this issue and (2) would make redirects etc. follow the same IP version override configs that the initial request has.

IMO this seems like the better option, I have a hacky version of this [implemented as a PoC](https://github.com/prometheus/blackbox_exporter/compare/master...jacksontj:issue_237?expand=1) -- if we're okay with this I'd be happy to make the changes it requires (1) a change to prometheus/common to have a function which allows for a base transport and (2) a change to the http probe (and presumably the other probes that do the same -- tcp/dns/icmp) to use this mechanism of overriding the dial mechanism.

**Comment by brian-brazil** [comment]
Thanks for looking into this, 2. was what I was generally thinking. That seems to have different DNS resolution semantics though, and I'm not sure the ServerName handling is correct on redirects.

**Comment by jacksontj** [comment]
What was the expected DNS resolution semantics? It looked like it's do 4
then 6 or 6 then 4. If so, this does that the only difference (right now)
is that it does that on any error.

As for servername, if you don't set servername in the config it works
(which is why I have the if block that sets if empty commented out). If you
want to specifically set a servername as is it will use that name always.
If we wanted to support explicit names and redirects we'd need #2 and we'd
need a list of servername, which seems like a much more significant change

**Comment by brian-brazil** [comment]
Yes, what you're doing is attempting a connection rather than just DNS.

> If we wanted to support explicit names and redirects we'd need #2 and we'd need a list of servername, which seems like a much more significant change

One of the expected use cases is that you'll be accessing a pile of replicas via their IP, and overriding the Host to get things to work. We should ensure that that'll work.

**Comment by jacksontj** [comment]
If you want to do the request by ip and set the host header and servername
then we have to make it so you can get a pointer to the tlsconfig so the
checkredirect method can update it.

I'm not sure why doing the DNS lookup first is better than just forcing the
request to a specific ip type? If there a particular reason (the code and
docs don't make that clear). Since the lookup is done immediately before
the request you should get the same behavior, the only difference being
that the URL remains the same (meaning default servername interpretation
works).

**Comment by jacksontj** [comment]
What might be an easier way to express the behavior is as if it were a series of shell commands

The current behavior (as I understand it):
```
dig AAAA <target> || dig A <target>
curl -H 'Host: targetname' <targetIP>:<targetPort>
```

The behaviour my hacky diff has:
```
curl -6 <targetIP>:<targetPort> || curl -4 <targetIP>:<targetPort>
```

AFAIK thats the only difference, and assuming there is a way to check if an err was a DNS error I could replicate the current behavior as well (by only retrying on 4 if the 6 failure was DNS).

> One of the expected use cases is that you'll be accessing a pile of replicas via their IP, and overriding the Host to get things to work. We should ensure that that'll work.

I assume that by this you mean you want to configure it with a bunch of IPs? If so that would work assuming the IP was in the certname, if you needed a specific name in the ServerName then you'd have to define it.

One of the issues that exists in the current exporter (as well as my implementation for #2) is specifying servernames in redirect chains. In the case where you want to *specify* a ServerName in the config its the same one always. IMO the ServerName option should be a list then the redirects could move the TLS ServerName up to the next one on the list.

**Comment by brian-brazil** [comment]
> AFAIK thats the only difference, and assuming there is a way to check if an err was a DNS error I could replicate the current behavior as well (by only retrying on 4 if the 6 failure was DNS).

That'd be using different code to handle DNS in two places, hooking in what's already there would be preferable.

> If so that would work assuming the IP was in the certname, if you needed a specific name in the ServerName then you'd have to define it.

The IP wouldn't be in the certname. Think probing a pile of backends of a load balancer to check that each works.

> In the case where you want to _specify_ a ServerName in the config its the same one always. IMO the ServerName option should be a list then the redirects could move the TLS ServerName up to the next one on the list.

It can't be a list as the library for that doesn't support that, but maybe when we stop specifying it'll work in enough scenarios to be okay.

**Comment by jacksontj** [comment]
> That'd be using different code to handle DNS in two places, hooking in what's already there would be preferable.

If thats the only concern I'd be happy to change the equivalent code in the other probes-- if this is the preferred mechanism.

> The IP wouldn't be in the certname. Think probing a pile of backends of a load balancer to check that each works.

Thats what I figured, just wanted to double check :)

> It can't be a list as the library for that doesn't support that

My suggestion is to use a list in combination with the `CheckRedirect` function. If we don't need to be explicit about setting the ServerName then this isn't necessary. And if we want to add that later we definitely can, it sounds like there isn't immediate interest in having that list-- so IMO we delay that until we need it (as it is a larger change).

> but maybe when we stop specifying it'll work in enough scenarios to be okay.

With this dial override approach we don't have to specify the ServerName as it uses the url Hostname (as you'd expect) so you get the same behavior without having to set the ServerName field.

If the only concern with my #2 approach is splitting the code for the probes, I can definitely do the same dial override for the others as well. Sounds like thats what we want?

**Comment by brian-brazil** [comment]
> If thats the only concern I'd be happy to change the equivalent code in the other probes-- if this is the preferred mechanism.

How would that work for all the metrics etc.? I can't see this working with ICMP, or UDP if we ever get it.

**Comment by jacksontj** [comment]
For both we'd just need to use the same [net.Dial](https://golang.org/pkg/net/#Dial) as it supports specifying tcp/udp/icmp as well as ip version.

**Comment by brian-brazil** [comment]
I see your point. How would you get the DNS resolution metric?

**Comment by jacksontj** [comment]
Depends on which metrics we get. If its just resolution time, adding a timer to https://golang.org/pkg/net/#Resolver Dial would do it, if we need to know where it connected to then adding a Control func to https://golang.org/pkg/net/#Dialer would do it.

**Comment by brian-brazil** [comment]
> If its just resolution time, adding a timer to https://golang.org/pkg/net/#Resolver Dial would do it

The docs indicate that Dial only creates the connection, which takes 0 time for UDP and doesn't include the actual query/response time.

**Comment by jacksontj** [comment]
True, so the only way to make that work would be to wrap the returned `net.Conn`. 

The only alternative that I see is changing the prometheus/common/config stuff to return a pointer to the TLSConfig such that we can modify it during our execution. This way we could continue to do the resolution separate and set the ServerName based on the actual URL that we want to hit.

**Comment by brian-brazil** [comment]
Do you want to do up a PR and we'll see how it looks?

I don't think it'd be sane to change the common config that way.

**Comment by discordianfish** [comment]
Just ran into this as well. Thanks to httpbin you can nicely reproduce this, e.g:

```
# wget -qO- 'http://localhost:9115/probe?target=https://httpbin.org/redirect-to?url=https%3A%2F%2Fgoogle.com&status_code=302&d
ebug=true&module=http_non5xx'
Logs for the probe:
ts=2019-02-25T23:24:01.498672935Z caller=main.go:118 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Beginning probe" probe=http timeout_seconds=9.5
ts=2019-02-25T23:24:01.498767502Z caller=utils.go:42 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Resolving target address" ip_protocol=ip4
ts=2019-02-25T23:24:01.528111205Z caller=utils.go:70 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Resolved target address" ip=34.197.95.106
ts=2019-02-25T23:24:01.528156148Z caller=http.go:349 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Making HTTP request" url="https://[34.197.95.106]/redirect-to?url=https://google.com" host=httpbin.org
ts=2019-02-25T23:24:01.545007923Z caller=http.go:305 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Received redirect" url=https://google.com
ts=2019-02-25T23:24:01.563605503Z caller=http.go:364 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=error msg="Error for HTTP request" err="Get https://google.com: x509: certificate is valid for *.google.com, *.android.com, *.appengine.google.com, *.cloud.google.com, *.g.co, *.gcp.gvt2.com, *.ggpht.cn, *.google-analytics.com, *.google.ca, *.google.cl, *.google.co.in, *.google.co.jp, *.google.co.uk, *.google.com.ar, *.google.com.au, *.google.com.br, *.google.com.co, *.google.com.mx, *.google.com.tr, *.google.com.vn, *.google.de, *.google.es, *.google.fr, *.google.hu, *.google.it, *.google.nl, *.google.pl, *.google.pt, *.googleadapis.com, *.googleapis.cn, *.googlecommerce.com, *.googlevideo.com, *.gstatic.cn, *.gstatic.com, *.gstaticcnapps.cn, *.gvt1.com, *.gvt2.com, *.metric.gstatic.com, *.urchin.com, *.url.google.com, *.youtube-nocookie.com, *.youtube.com, *.youtubeeducation.com, *.youtubekids.com, *.yt.be, *.ytimg.com, android.clients.google.com, android.com, developer.android.google.cn, developers.android.google.cn, g.co, ggpht.cn, goo.gl, google-analytics.com, google.com, googlecommerce.com, source.android.google.cn, urchin.com, www.goo.gl, youtu.be, youtube.com, youtubeeducation.com, youtubekids.com, yt.be, not httpbin.org"
ts=2019-02-25T23:24:01.563644754Z caller=http.go:450 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Response timings for roundtrip" roundtrip=0 start=2019-02-25T23:24:01.528231995Z dnsDone=2019-02-25T23:24:01.528231995Z connectDone=2019-02-25T23:24:01.529506981Z gotConn=2019-02-25T23:24:01.542526608Z responseStart=2019-02-25T23:24:01.544947141Z end=0001-01-01T00:00:00Z
ts=2019-02-25T23:24:01.563669017Z caller=http.go:450 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=info msg="Response timings for roundtrip" roundtrip=1 start=2019-02-25T23:24:01.54509146Z dnsDone=2019-02-25T23:24:01.552169426Z connectDone=2019-02-25T23:24:01.553488627Z gotConn=0001-01-01T00:00:00Z responseStart=0001-01-01T00:00:00Z end=0001-01-01T00:00:00Z
ts=2019-02-25T23:24:01.563686653Z caller=main.go:131 module=http_non5xx target="https://httpbin.org/redirect-to?url=https://google.com" level=error msg="Probe failed" duration_seconds=0.064972577
```

**Comment by jacksontj** [comment]
@brian-brazil https://github.com/prometheus/blackbox_exporter/compare/master...jacksontj:issue_237_2?expand=1 a hacky implementation of the CheckRedirect option.

**Comment by brian-brazil** [comment]
I meant the wrapping of net.Conn. The CheckRedirect has the issue that it'd break redirects to subsequent hosts.

**Comment by brian-brazil** [comment]
That's a timeout, and thus unrelated. If you believe there's a bug, please file a new one rather than posting on a closed one.
